libs/jwst-storage/src/docs/storage.rs

```diff
@@ -204,14 +246,109 @@ impl DocStorage for ORM {
     }
 }
 
+#[async_trait]
+impl DocSync for DocAutoStorage {
+    async fn sync(&self, id: String, remote: String) -> io::Result<()> {
+        if let Entry::Vacant(entry) = self.remote.entry(id.clone()) {
+            let workspace = self.get(id).await.map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::BrokenPipe,
+                    format!("Failed to get doc: {}", e),
+                )
+            })?;
+
+            let (tx, mut rx) = channel(100);
+
+            debug!("spawn sync thread");
+            std::thread::spawn(move || {
+                let rt = tokio::runtime::Runtime::new().expect("Failed to create runtime");
+                rt.block_on(async move {
+                    debug!("generate remote config");
+                    let mut req = Url::parse(&remote)
+                        .expect( "Failed to parse remote url")
+                        .into_client_request()
+                        .expect("Failed to create client request");
+                    req.headers_mut()
+                        .append("Sec-WebSocket-Protocol", HeaderValue::from_static("AFFiNE"));
+
+                    debug!("connect to remote: {}", req.uri());
+                    let (socket, _) = connect_async(req)
+                        .await.expect("Failed to connect to remote");
+                    let (mut socket_tx, mut socket_rx) = socket.split();
+
+                    debug!("sync init message");
+                    match workspace.read().await.sync_init_message() {
+                        Ok(init_data) => {
+                            debug!("send init message");
+                            if let Err(e) = socket_tx.send(Message::Binary(init_data)).await {
+                                error!("Failed to send init message: {}", e);
+                                return;
+                            }
+                        }
+                        Err(e) => {
+                            error!("Failed to create init message: {}", e);
+                            return;
+                        }
+                    }
+
+                    let id = workspace.read().await.id();
+                    debug!("start sync thread {id}");
+                    loop {
+                        tokio::select! {
```

Comment:
sync to remote yclient, client can refer cc @linonetwo

Reply:
note that the logic has extra wrapping (refer to sync_decode_message), exchange update to single source with local ipc may not need these wrapping.